File: e2e/pages/LoginPage.js
import { expect } from '@playwright/test';

export class LoginPage {
  /**
   * @param {import('@playwright/test').Page} page
   */
  constructor(page) {
    this.page = page;
    this.emailInput = page.locator('input[type="email"]');
    this.passwordInput = page.locator('input[type="password"]');
    this.loginButton = page.locator('button[type="submit"]');
    this.googleLoginButton = page.locator('button', { hasText: /Google/i });
    this.errorMessage = page.getByRole('status');
  }

  async goto() {
    await this.page.goto('/login');
  }

  async login(email, password, { waitForDashboard = true } = {}) {
    await this.emailInput.fill(email);
    await this.passwordInput.fill(password);
    await this.loginButton.click();
    if (waitForDashboard) {
      await expect(this.page).toHaveURL('/dashboard', { timeout: 10000 });
    }
  }

  async getErrorMessage() {
    const status = this.errorMessage;
    try {
      await expect(status).toBeVisible({ timeout: 5000 });
      const text = await status.textContent();
      return text?.trim() || null;
    } catch {
      return null;
    }
  }
}

File: selenium/src/main/java/pages/LoginPage.java
package pages;

import base.BasePage;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;

public class LoginPage extends BasePage {

    // Locators
    private final By emailInput = By.cssSelector("input[type='email']");
    private final By passwordInput = By.cssSelector("input[type='password']");
    private final By loginButton = By.cssSelector("button[type='submit']");
    private final By errorMessage = By.cssSelector("[role='status'], [role='alert'], .text-red-500");
    private final By googleLoginButton = By.xpath("//button[contains(., 'Google')]");

    public LoginPage(WebDriver driver) {
        super(driver);
    }

    public void login(String email, String password) {
        type(emailInput, email);
        type(passwordInput, password);
        click(loginButton);
    }

    public String getErrorMessage() {
        if (isDisplayed(errorMessage)) {
            return getText(errorMessage);
        }
        return null;
    }

    public boolean isGoogleLoginButtonVisible() {
        return isDisplayed(googleLoginButton);
    }
}
